contributors/design-proposals/termination-reason.md

```diff
@@ -0,0 +1,345 @@
+The logic associated with container cleanup will be executed by the container's
+entry point during the container's configured termination grace period.
+
+## Requirements
+
+1. Termination reason must be available to a container at the time its pre-stop 
+handler is invoked. In this way, applications can perform any actions
+necessary to configure the container's entry point prior to the entry point 
+receiving the `TERM` signal.
+1. The delivery mechanism for termination reason must be consumable by, and 
+ compatible with, existing HTTP and command based pre-stop handlers.
+1. Any headers, variables, or parameters used to communicate a termination 
+reason must not be hard coded, and the receiving container must be able to 
+override the defaults.
+
+## API Objects
+
+Termination reason is implemented by a string type.
+
+```golang
+// TerminationReason is a string type used to indicate the reason for 
+// termination to the preStop lifecycle hook of a Pod's containers.
+type TerminationReason string
+
+const (
+	// ReasonEviction is the default reason used to communicate that a Pod has 
+ 	// been terminated due to an eviction.
+	ReasonEviction TerminationReason = "Eviction"
+	// ReasonIntolerableTaint is the default reason used to communicate that a Pod 
+	// has been terminated due to a taint for which it has no declared toleration.
+	ReasonIntolerableTaint = "IntolerableTaint"
+	// ReasonDecommissioned is the default reason sent by controllers to 
+	// indicate that a Pod has been terminated due to a horizontal scaling 
+	// event. Pods receiving this reason should not expect to be rescheduled.
+	ReasonDecommissioned = "Decommissioned"
+	// ReasonUpdate is the reason sent by a controller or User to indicate that 
+	// a Pod has been terminated for the purposes of a destructive update. Pods 
+	// receiving this signal should expect to be rescheduled immediately. Note
+	// that this DOES NOT imply that scheduling will succeed.
+	ReasonUpdate = "Update"
+)
+
+```
+
+DeleteOptions is modified to carry the termination reason.
+
+```golang
+type DeleteOptions struct {
+	// Other fields omitted for brevity.
+	
+	// Reason indicates the reason for the Pod's termination. This field may 
+	// be supplied by a user or a controller.
+	Reason *TerminationReason `json:"reason,omitempty"`
+}
+```
+
+The ObjectMeta struct is modified to carry the termination reason via 
+a Pod's Metadata.
+
+```golang
+type ObjectMeta {
+	// Other fields omitted for brevity.
+	
+        // TerminationReason indicates the reason for the Pod's termination.
+        // This field may be supplied by a user or a controller.
+	TerminationReason *TerminationReason `json:"reason,omitempty"`
+}
+```
+
+TerminationReasonDelivery provides configuration delivery method of Lifecycle's 
+`PreStop` Handler.
+
+```golang
+// TerminationReasonDelivery is used to configure the delivery method for
+// termination reasons. It is a union type, and exactly one of the fields may be
```

Comment:
Do you mean by extending HttpAction and CommandAction to contain the appropriate fields? If so, I think we'd have to consider that these fields are only valid in the context of a pre-stop hook. We could modify validation to ensure that they are only used in this context and not in Probe's Handlers, but I'm not convinced that is cleaner.